design/providerid_format_change.md

```diff
@@ -0,0 +1,154 @@
+cluster API to fail, blocking the upgrade.
+
+It could be made to be dynamic by including other resources that are generated
+uniquely for each kubernetes deployment.
+
+The current providerID logic makes use of labels in
+`metal3.io/uuid=<bmh-uuid>` format, also supplied via templates. This does not
+provider extra information as provider-id (`metal3://<bmh-uuid>`) contains the
+same information. In this regard, the label is redundant.
+
+### Goals
+
+- Add new implicit field to `metal3datatemplate`
+- Change providerID format to include dynamic information
+- Deprecate usage of node labels.
+
+### Non-Goals
+
+Making tempates simpler.
+
+## Proposal
+
+The issues discussed above can be solved by introducing a new field in the
+metal3datatemplate implicitly. This does not require a change in the API
+(metal3datatemplate object) if CAPM3 always automatically populates this key in
+the resulting secret for the node. But it should be a default only, so that it
+could be overwritten by the user if they define this key explicitly.
+
+```bash
+providerid: "metal3://<namespace>/<bmh name>/<metal3machine name>"
+```
+
+The new format is both dynamic and known in advance. Since this is added from
+the code, there is no need of changing Spec or Status sections of the resource.
+
+Also, since it is constructed in the code, there is no need for the users to
+provider provider-id and label via templates. However, the user still refer to
+the providerID variable in their templates to get it populated by cloud-init
+
+### User Stories
+
+- As a user, I want my nodes' providerIDs to be unique per kubernetes deployment
+ and be generated automatically by CAPM3 using metal3machine and baremetalhost resource names.
+
+## Design Details
+
+The following changes are required in the code:
+
+- Searching nodes is no longer done using labels
+- Searching nodes is done using the implicitly set providerid
+- Providerid field is introduced in metal3datatemplate in the code.
+- Labels are no longer provided via templates
+
+Concerning backward compatibility, there are three scenarios
+
+1. backward compatibility with legacy providerID existing deployments,
+  - During scaling out, new nodes keep using legacy providerID
+  - During upgrading, upgraded nodes keep using legacy providerID
+2. backwards compatibility with legacy providerID for new deployments,
+  - During provisioning, if user supplies template with legacy providerID and
+  node label, then the nodes should be provisioned with the legacy providerID
+  - This will be supported for the next two releases
+
+3. transition from legacy to new providerID format for existing deployments.
+  - Existing nodes with legacy providerID can be moved to the new format using
+  upgrade. This can be achieved by creating new
+  `kubeadmconfigs.bootstrap.cluster.x-k8s.io` resource with the new provider-id
+  set in side the `.spec.joinConfiguration.nodeRegistration.kubeletExtraArgs`
```

Comment:
I would really appreciate an example here.

Reply:
Thanks. Added an example. 
also addresses the other comments.